File: public/exports/spec/06_scripts/signature-demo/verify_reference.py
# ===== verify_reference.py =====
#!/usr/bin/env python3
"""
LLMFeed Reference Verification Script
Simple, clean reference implementation for verifying LLMFeed signatures

Usage: python verify_reference.py signed.llmfeed.json
"""

import json
import sys
import base64
import requests
from pathlib import Path
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
from cryptography.exceptions import InvalidSignature

def load_public_key_from_url(url):
    """Load Ed25519 public key from URL"""
    try:
        response = requests.get(url, timeout=10)
        response.raise_for_status()
        
        public_key = serialization.load_pem_public_key(response.content)
        
        if not isinstance(public_key, Ed25519PublicKey):
            raise ValueError(f"Key must be Ed25519, got {type(public_key).__name__}")
        
        return public_key
    except Exception as e:
        raise RuntimeError(f"Failed to load public key from {url}: {e}")

def llmfeed_canonicalize(data):
    """
    LLMFeed canonical JSON serialization
    CRITICAL: Preserves key order for LLM semantics
    """
    return json.dumps(data, separators=(',', ':'), ensure_ascii=False).encode('utf-8')

def verify_feed(feed_path):
    """Verify a signed LLMFeed file"""
    
    # Load feed
    with open(feed_path, 'r', encoding='utf-8') as f:
        feed = json.load(f)
    
    # Validate structure
    if 'trust' not in feed:
        raise ValueError("Feed missing 'trust' block")
    if 'signature' not in feed:
        raise ValueError("Feed missing 'signature' block")
    
    trust = feed['trust']
    signature_block = feed['signature']
    
    # Validate trust block
    required_trust_fields = ['signed_blocks', 'algorithm', 'canonicalization', 'public_key_hint']
    for field in required_trust_fields:
        if field not in trust:
            raise ValueError(f"Trust block missing '{field}' field")
    
    # Validate algorithm
    if trust['algorithm'] != 'ed25519':
        raise ValueError(f"Unsupported algorithm: {trust['algorithm']}")
    
    # Validate canonicalization
    if trust['canonicalization'] != 'https://llmca.org/mcp-canonical-json/v1':
        raise ValueError(f"Unsupported canonicalization: {trust['canonicalization']}")
    
    # Validate signature block
    if 'value' not in signature_block:
        raise ValueError("Signature block missing 'value' field")
    
    # Get signed blocks
    signed_blocks = trust['signed_blocks']
    
    # Build partial feed
    partial_feed = {}
    for block_name in signed_blocks:
        if block_name not in feed:
            raise ValueError(f"Missing signed block: {block_name}")
        partial_feed[block_name] = feed[block_name]
    
    # Canonicalize (CRITICAL: preserves key order)
    payload = llmfeed_canonicalize(partial_feed)
    
    # Load public key
    public_key = load_public_key_from_url(trust['public_key_hint'])
    
    # Decode signature
    try:
        signature_bytes = base64.b64decode(signature_block['value'])
    except Exception as e:
        raise ValueError(f"Invalid base64 signature: {e}")
    
    # Verify signature
    try:
        public_key.verify(signature_bytes, payload)
        return True
    except InvalidSignature:
        return False

def main():
    if len(sys.argv) != 2:
        print("Usage: python verify_reference.py signed.llmfeed.json")
        sys.exit(1)
    
    feed_path = sys.argv[1]
    
    # Validate arguments
    if not Path(feed_path).exists():
        print(f"❌ Feed file not found: {feed_path}")
        sys.exit(1)
    
    try:
        print(f"🔍 Verifying feed: {feed_path}")
        
        # Load feed to show info
        with open(feed_path, 'r', encoding='utf-8') as f:
            feed = json.load(f)
        
        trust = feed.get('trust', {})
        print(f"   Algorithm: {trust.get('algorithm', 'unknown')}")
        print(f"   Signed blocks: {trust.get('signed_blocks', [])}")
        print(f"   Public key: {trust.get('public_key_hint', 'unknown')}")
        
        # Verify
        is_valid = verify_feed(feed_path)
        
        if is_valid:
            print("✅ Signature is VALID")
            sys.exit(0)
        else:
            print("❌ Signature is INVALID")
            sys.exit(1)
            
    except Exception as e:
        print(f"❌ Verification failed: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: public/exports/spec/06_scripts/signature-demo/verify_reference.js
# ===== verify_reference.js =====
#!/usr/bin/env node
/**
 * LLMFeed Reference Verification Script (JavaScript)
 * Simple, clean reference implementation for verifying LLMFeed signatures
 * 
 * Usage: node verify_reference.js signed.llmfeed.json
 */

const fs = require('fs');
const crypto = require('crypto');
const https = require('https');

function loadPublicKeyFromUrl(url) {
    /**
     * Load Ed25519 public key from URL
     */
    return new Promise((resolve, reject) => {
        const request = https.get(url, { timeout: 10000 }, (response) => {
            let data = '';
            
            response.on('data', (chunk) => {
                data += chunk;
            });
            
            response.on('end', () => {
                try {
                    const publicKey = crypto.createPublicKey(data);
                    
                    if (publicKey.asymmetricKeyType !== 'ed25519') {
                        reject(new Error(`Key must be Ed25519, got ${publicKey.asymmetricKeyType}`));
                        return;
                    }
                    
                    resolve(publicKey);
                } catch (error) {
                    reject(new Error(`Failed to parse public key: ${error.message}`));
                }
            });
        });
        
        request.on('error', (error) => {
            reject(new Error(`Failed to load public key from ${url}: ${error.message}`));
        });
        
        request.on('timeout', () => {
            request.destroy();
            reject(new Error(`Timeout loading public key from ${url}`));
        });
    });
}

function llmfeedCanonicalize(data) {
    /**
     * LLMFeed canonical JSON serialization
     * CRITICAL: Preserves key order for LLM semantics
     */
    return JSON.stringify(data, null, 0);
}

async function verifyFeed(feedPath) {
    /**
     * Verify a signed LLMFeed file
     */
    
    // Load feed
    const feedData = fs.readFileSync(feedPath, 'utf-8');
    const feed = JSON.parse(feedData);
    
    // Validate structure
    if (!feed.trust) {
        throw new Error("Feed missing 'trust' block");
    }
    if (!feed.signature) {
        throw new Error("Feed missing 'signature' block");
    }
    
    const trust = feed.trust;
    const signatureBlock = feed.signature;
    
    // Validate trust block
    const requiredTrustFields = ['signed_blocks', 'algorithm', 'canonicalization', 'public_key_hint'];
    for (const field of requiredTrustFields) {
        if (!trust[field]) {
            throw new Error(`Trust block missing '${field}' field`);
        }
    }
    
    // Validate algorithm
    if (trust.algorithm !== 'ed25519') {
        throw new Error(`Unsupported algorithm: ${trust.algorithm}`);
    }
    
    // Validate canonicalization
    if (trust.canonicalization !== 'https://llmca.org/mcp-canonical-json/v1') {
        throw new Error(`Unsupported canonicalization: ${trust.canonicalization}`);
    }
    
    // Validate signature block
    if (!signatureBlock.value) {
        throw new Error("Signature block missing 'value' field");
    }
    
    // Get signed blocks
    const signedBlocks = trust.signed_blocks;
    
    // Build partial feed
    const partialFeed = {};
    for (const blockName of signedBlocks) {
        if (feed[blockName] === undefined) {
            throw new Error(`Missing signed block: ${blockName}`);
        }
        partialFeed[blockName] = feed[blockName];
    }
    
    // Canonicalize (CRITICAL: preserves key order)
    const payload = Buffer.from(llmfeedCanonicalize(partialFeed), 'utf-8');
    
    // Load public key
    const publicKey = await loadPublicKeyFromUrl(trust.public_key_hint);
    
    // Decode signature
    let signatureBytes;
    try {
        signatureBytes = Buffer.from(signatureBlock.value, 'base64');
    } catch (error) {
        throw new Error(`Invalid base64 signature: ${error.message}`);
    }
    
    // Verify signature
    try {
        return crypto.verify(null, payload, publicKey, signatureBytes);
    } catch (error) {
        throw new Error(`Signature verification failed: ${error.message}`);
    }
}

async function main() {
    const args = process.argv.slice(2);
    
    if (args.length !== 1) {
        console.log("Usage: node verify_reference.js signed.llmfeed.json");
        process.exit(1);
    }
    
    const feedPath = args[0];
    
    // Validate arguments
    if (!fs.existsSync(feedPath)) {
        console.log(`❌ Feed file not found: ${feedPath}`);
        process.exit(1);
    }
    
    try {
        console.log(`🔍 Verifying feed: ${feedPath}`);
        
        // Load feed to show info
        const feedData = fs.readFileSync(feedPath, 'utf-8');
        const feed = JSON.parse(feedData);
        
        const trust = feed.trust || {};
        console.log(`   Algorithm: ${trust.algorithm || 'unknown'}`);
        console.log(`   Signed blocks: ${(trust.signed_blocks || []).join(', ')}`);
        console.log(`   Public key: ${trust.public_key_hint || 'unknown'}`);
        
        // Verify
        const isValid = await verifyFeed(feedPath);
        
        if (isValid) {
            console.log("✅ Signature is VALID");
            process.exit(0);
        } else {
            console.log("❌ Signature is INVALID");
            process.exit(1);
        }
        
    } catch (error) {
        console.log(`❌ Verification failed: ${error.message}`);
        process.exit(1);
    }
}

if (require.main === module) {
    main();
}